File: SequenceTransformation/cpp/st.cpp
// sequence transformation soln in cpp for codeeval by steven a dunn

#include <fstream>
#include <iostream>
#include <regex>
#include <sstream>
#include <string>
#include <vector>

using std::cout;
using std::endl;
using std::ifstream;
using std::regex;
using std::regex_match;
using std::string;
using std::stringstream;
using std::vector;

string is_sequence_transformation(string, string);
string stringify(vector<string>);

string stringify(vector<string> vec)
{
    stringstream ss;
    for (int i = 0; i < vec.size(); ++i)
    {
        ss << vec[i];
    }
    return ss.str();
}

string is_sequence_transformation(string binary, string text)
{
    const string ZERO_REGEX = "A+";
    const string ONE_REGEX = "(A+|B+)";
    vector<string> regex_vec;

    for (int binary_idx = 0; binary_idx < binary.length(); ++binary_idx)
    {
        (binary[binary_idx] == '0') ? regex_vec.push_back(ZERO_REGEX) : regex_vec.push_back(ONE_REGEX);
    }

    string text_regex = stringify(regex_vec);

    regex reg(text_regex);
    string result;
    regex_match(text, reg) ? result = "Yes" : result = "No";

    return result;
}

int main(int argc, char* argv[])
{
    ifstream inputFile(argv[1]);
    string line;

    if (inputFile)
    {
        while (getline(inputFile, line))
        {
            string binary = line.substr(0, line.find(" "));
            string text = line.substr(line.find(" ") + 1);
            cout << is_sequence_transformation(binary, text) << endl;
        }
        inputFile.close();
    }
    return 0;
}

File: SequenceTransformation/rb/st.rb

ZERO_REGEX = "A+"
ONE_REGEX = "(A+|B+)"

File.open(ARGV[0], 'r').each_line do |line|
    binary, text = line.chomp.split
    text_regex = []
    binary.split("").each do |digit|
        digit == "0" ? text_regex << ZERO_REGEX : text_regex << ONE_REGEX
    end
    regex = Regexp.new(text_regex.join)
    result = text =~ regex ? "Yes" : "No"
    puts result
end


=begin
File.open(ARGV[0], 'r').each_line do |line|
    binary, text = line.chomp.split(" ")

    text = text.squeeze('A-B')

    perms = []

    binary.split("").each do |digit|
        if digit == "1"
            if perms.empty?
                perms << "A"
                perms << "B"
            else
                perms_a = perms.map{ |x| x += "A" }
                perms_b = perms.map{ |x| x += "B" }
                perms = perms_a.concat(perms_b)
                perms.map! { |x| x.squeeze('A-B') }
            end
        else
            if perms.empty?
                perms << "A"
            end
            perms.map!{ |x| x += "A" }
            perms.map! { |x| x.squeeze('A-B') }
        end

        perms.each do |perm|
            if perm != text[0...perm.length]
                perms.delete(perm)
            end
        end
    end

    if perms.include? text
        puts "Yes"
    else
        puts "No"
    end
end
=end